core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala

```diff
@@ -74,6 +74,10 @@ class LocalSparkCluster(
 
   def stop(): Unit = {
     logInfo("Shutting down local Spark cluster.")
+    // SPARK-31922: wait one more second before shutting down rpcEnvs of master and worker,
+    // in order to let the cluster have time to handle the `UnregisterApplication` message.
+    // Otherwise, we could hit "RpcEnv already stopped" error.
+    Thread.sleep(1000)
```

Comment:
`1000ms` is from my test result. Actually, 500ms could also work. I just want to be conservative without an obvious pause when exiting spark-shell.

Reply:
I don't think this is ideal to handle the issue, there can always be lags for various reasons so the error message may still appear. Ideally we should get to know the LocalCluster is stopped on the Master side and don't output the RpcEnv stopped error message but I'm not sure how hard it is to do so.

Reply:
+1, can we have unregister happen synchronously for the local cluster https://github.com/Ngone51/spark/pull/3

Reply:
Hi @gerashegalov , thank you for the PR. But I think the PR might not work. Sync from AppClient to Master is not enough. We should also sync the events from Master to Worker after the Master receives the `UnregisterApplication` message. In `Master. finishApplication`, you can see that the Master not only reply to AppClient, but also send messages to Worker in an async way. And the error happens exactly on the Worker. But if we do in this way, I'm afraid we add too much special code logic for local cluster only, which I personally think can be overhead.

Reply:
thanks for taking a look @Ngone51 . It's just a sample but, on my box the issue has not shown up after the PR making sure the client is stopped only after finishApplication and reworking the shutdown logic. Agree that perfecting it comes at a price for minimum returns given it's a corner case.